ai_economist/foundation/components/redistribution.py

```diff
@@ -379,6 +429,390 @@ def set_new_period_rates_model(self):
+        # Until we reach the min sample number, keep checking if we have reached it
+        if not self._reached_min_samples:
+            # Note: self.saez_buffer includes the global buffer (if applicable)
+            if len(self.saez_buffer) >= self._buffer_size:
+                self._reached_min_samples = True
+
+        # If no enough samples, use random taxes.
+        if not self._reached_min_samples:
+            self.curr_bracket_tax_rates = np.random.uniform(
+                low=self.rate_min,
+                high=self.curr_rate_max,
+                size=self.curr_bracket_tax_rates.shape,
+            )
+            return
+
+        incomes_and_marginal_rates = np.array(self.saez_buffer)
+
+        # Elasticity assumed constant for all incomes
+        # (Run this for the sake of tracking the estimate; will not actually use the
+        # estimate if using fixed elasticity.)
+        if update_elas_tm1:
+            self.elas_tm1 = float(self.elas_t)
+        if update_log_z0_tm1:
+            self.log_z0_tm1 = float(self.log_z0_t)
+
+        elas_t, log_z0_t = self.estimate_uniform_income_elasticity(
+            incomes_and_marginal_rates,
+            elas_df=0.98,
+            elas_tm1=self.elas_tm1,
+            log_z0_tm1=self.log_z0_tm1,
+            verbose=False,
+        )
+
+        if update_elas_tm1:
+            self.elas_t = float(elas_t)
+        if update_log_z0_tm1:
+            self.log_z0_t = float(log_z0_t)
+
+        # If a fixed estimate has been specified, use it in the formulas below
+        if self._saez_fixed_elas is not None:
+            elas_t = float(self._saez_fixed_elas)
+
+        # Get Saez parameters at each income bin to compute a marginal tax rate schedule
+        binned_gzs, binned_azs = self.get_binned_saez_welfare_weight_and_pareto_params(
+            population_incomes=incomes_and_marginal_rates[:, 0]
+        )
+
+        # Use the elasticity to compute this binned schedule using the Saez formula
+        binned_marginal_tax_rates = self.get_saez_marginal_rates(
+            binned_gzs, binned_azs, elas_t
+        )
+
+        # Adapt the saez tax schedule to the tax brackets
+        self.curr_bracket_tax_rates = np.clip(
+            self.bracketize_schedule(
+                bin_marginal_rates=binned_marginal_tax_rates,
+                bin_edges=self._saez_income_bin_edges,
+                bin_sizes=self._saez_income_bin_sizes,
+            ),
+            self.rate_min,
+            self.curr_rate_max,
+        )
+
+        self.running_avg_tax_rates = (self.running_avg_tax_rates * 0.99) + (
+            self.curr_bracket_tax_rates * 0.01
+        )
+
+    # Implementation of the Saez formula in this periodic, bracketed setting
+    # ----------------------------------------------------------------------
+    @property
+    def saez_buffer(self):
+        if not self._global_saez_buffer:
+            return self._local_saez_buffer
+        if self._additions_this_episode == 0:
```

Comment:
use "if elif else"

Reply:
Same as above

Reply:
Used if elif for readability here as well